dev/system-test/firestore.ts

```diff
@@ -2321,3 +2324,271 @@ describe('Client initialization', () => {
+
+  it('succeeds with document snapshots', async () => {
+    const bundle = firestore.bundle('test-bundle');
+    const snap1 = await testCol.doc('doc1').get();
+    const snap2 = await testCol.doc('doc2').get();
+
+    bundle.add(snap1);
+    bundle.add(snap2);
+    // `elements` is expected to be [bundleMeta, doc1Meta, doc1Snap, doc2Meta, doc2Snap].
+    const [lengths, elements] = bundleToLengthsAndObject(
+      (await bundle.build()).toString()
+    );
+
+    expect(lengths.length)
+      .to.equal(elements.length)
+      .to.equal(5);
+
+    const meta = (elements[0] as IBundleElement).metadata;
+    expect(meta).to.deep.equal({
+      id: 'test-bundle',
+      createTime: snap2.readTime.toProto().timestampValue
+    });
+
+    // Verify doc1Meta and doc1Snap
+    const result1 = [
+      (elements[1] as IBundleElement).documentMetadata,
+      (elements[2] as IBundleElement).document,
+    ];
+    expect(result1).to.deep.equal([
+      {
+        documentKey: snap1.toDocumentProto().name,
+        readTime: snap1.readTime.toProto().timestampValue,
+      },
+      snap1.toDocumentProto(),
+    ]);
+
+    // Verify doc1Meta and doc1Snap
+    const result2 = [
+      (elements[3] as IBundleElement).documentMetadata,
+      (elements[4] as IBundleElement).document,
+    ];
+    expect(result2).to.deep.equal([
+      {
+        documentKey: snap2.toDocumentProto().name,
+        readTime: snap2.readTime.toProto().timestampValue,
+      },
+      snap2.toDocumentProto(),
+    ]);
+  });
+
+  it('succeeds when nothing is added', async () => {
+    const bundle = firestore.bundle('test-bundle');
+
+    // `elements` is expected to be [bundleMeta].
+    const [lengths, elements] = bundleToLengthsAndObject(
+        (await bundle.build()).toString()
+    );
+
+    expect(lengths.length)
+        .to.equal(elements.length)
+        .to.equal(1);
+
+    const meta = (elements[0] as IBundleElement).metadata;
+    expect(meta).to.deep.equal({
+      id: 'test-bundle',
+      createTime: new Timestamp(0,0).toProto().timestampValue
+    });
+  });
+
+  it('succeeds when there are no results', async () => {
+    const bundle = firestore.bundle('test-bundle');
+    const query = testCol.where('sort', '==', 5);
+    const snap = await query.get();
+
+    bundle.add('query', snap);
+    // `elements` is expected to be [bundleMeta, query].
+    const [lengths, elements] = bundleToLengthsAndObject(
+      (await bundle.build()).toString()
+    );
+
+    expect(lengths.length)
+      .to.equal(elements.length)
+      .to.equal(2);
+
+    const meta = (elements[0] as IBundleElement).metadata;
+    expect(meta).to.deep.equal({
+      id: 'test-bundle',
+      createTime: snap.readTime.toProto().timestampValue
+    });
+
+    const namedQuery = (elements[1] as IBundleElement).namedQuery;
+    // Verify saved query.
+    expect(namedQuery).to.deep.equal({
+      name: 'query',
+      readTime: snap.readTime.toProto().timestampValue,
+      bundledQuery: extend(
```

Comment:
Nit: Replace all usages of `extend` with `{ ...oldData }`.

Reply:
There are other fields: `select` and `offset`, besides cursors. I'd prefer to fixing this in a separate PR.

Reply:
Works for me.